### ISSUE TITLE
Relax total-function check

### ISSUE THREAD

**Issue by ttuegel** [kframework/kore#2305]
We must move [this check](https://github.com/kframework/kore/blob/c160ec57e3c9c1c371493452f493f3323a7bf38c/kore/src/Kore/Step/Function/Evaluator.hs#L100) to relax its condition. Currently, it fails when a total function is given `\bottom` as an argument, which should not be an error.

**Comment by ttuegel** [comment]
Here is another [example](https://github.com/kframework/kore/pull/2350/checks?check_run_id=1705108061) of a test which should pass, but actually fails:

```
↓ test/Driver.hs
↓ Test
↓ Kore
↓ Step
↓ Simplification
↓ IntegrationProperty
✗ simplify returns simplified pattern

  ✗ simplify returns simplified pattern failed at test/Test/SMT.hs:43:24
    after 14 tests.
  
       ┏━━ test/Test/SMT.hs ━━━
    37 ┃ testPropertyWithoutSolver
    38 ┃     :: HasCallStack
    39 ┃     => String
    40 ┃     -> PropertyT NoSMT ()
    41 ┃     -> TestTree
    42 ┃ testPropertyWithoutSolver str =
    43 ┃     testProperty str . Hedgehog.property . Morph.hoist runNoSMT
       ┃     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       ┃     │ ━━━ Exception (SomeEntry) ━━━
       ┃     │ Evaluating total function
       ┃     │     /* Spa */
       ┃     │     elementList{}(
       ┃     │         /* Sfa */
       ┃     │         /* evaluated: */
       ┃     │         /* Spa */
       ┃     │         functionalConstr30{}(
       ┃     │             /* D Sfa Cl */ constr00{}(),
       ┃     │             /* Spa */ plain10{}(/* Fn Sfa */ \bottom{testSort{}}()),
       ┃     │             /* Fl Fn D Sfa Cl */ d{}()
       ┃     │         )
       ┃     │     )
       ┃     │ has resulted in \bottom.
  
    This failure can be reproduced by running:
    > recheck (Size 13) (Seed 11012454790019588701 2278687502285288851) simplify returns simplified pattern
  
Use '--hedgehog-replay "Size 13 Seed 11012454790019588701 2278687502285288851"' to reproduce.
```

**Comment by MirceaS** [comment]
@ttuegel I can't reproduce this. The test passes on my local machine every time I run it on both `master` and `entr`...

**Comment by ttuegel** [comment]
Also this pattern:

```
       ┏━━ test/Test/SMT.hs ━━━
    37 ┃ testPropertyWithoutSolver
    38 ┃     :: HasCallStack
    39 ┃     => String
    40 ┃     -> PropertyT NoSMT ()
    41 ┃     -> TestTree
    42 ┃ testPropertyWithoutSolver str =
    43 ┃     testProperty str . Hedgehog.property . Morph.hoist runNoSMT
       ┃     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       ┃     │ ━━━ Exception (SomeEntry) ━━━
       ┃     │ Evaluating total function
       ┃     │     /* Spa */
       ┃     │     concatList{}(
       ┃     │         /* Sfa */
       ┃     │         /* evaluated: */
       ┃     │         /*  */
       ┃     │         \and{listSort{}}(
       ┃     │             /* D Sfa */ \top{listSort{}}(),
       ┃     │             /*  */
       ┃     │             /* InternalList: */ concatList{}(
       ┃     │                 elementList{}(
       ┃     │                     /*  */
       ┃     │                     plain11{}(
       ┃     │                         /*  */
       ┃     │                         plain20{}(
       ┃     │                             /*  */
       ┃     │                             function20MapTest{}(
       ┃     │                                 /*  */
       ┃     │                                 \nu{}(
       ┃     │                                     Sv:mapSort{},
       ┃     │                                     /*  */
       ┃     │                                     \exists{mapSort{}}(
       ┃     │                                         f:testSort{},
       ┃     │                                         /* Fl Fn D */ unitMap{}()
       ┃     │                                     )
       ┃     │                                 ),
       ┃     │                                 /* D Sfa */ \top{testSort{}}()
       ┃     │                             ),
       ┃     │                             /*  */
       ┃     │                             functionalConstr12{}(
       ┃     │                                 /*  */
       ┃     │                                 \exists{testSort{}}(
       ┃     │                                     s:testSort1{},
       ┃     │                                     /* Fn */ h{}(/* Fn */ cf{}())
       ┃     │                                 )
       ┃     │                             )
       ┃     │                         )
       ┃     │                     )
       ┃     │                 ),
       ┃     │             concatList{}(
       ┃     │                 elementList{}(
       ┃     │                     /*  */
       ┃     │                     functionalSMT{}(
       ┃     │                         /*  */
       ┃     │                         constr20{}(
       ┃     │                             /*  */
       ┃     │                             injective11{}(
       ┃     │                                 /*  */
       ┃     │                                 plain10{}(
       ┃     │                                     /* Fn */
       ┃     │                                     functionalConstr12{}(/* Fn */ cg{}())
       ┃     │                                 )
       ┃     │                             ),
       ┃     │                             /* Fl Fn D */ functional00{}()
       ┃     │                         )
       ┃     │                     )
       ┃     │                 ),
       ┃     │             concatList{}(
       ┃     │                 elementList{}(
       ┃     │                     /* Spa */
       ┃     │                     functionalConstr11{}(
       ┃     │                         /* Spa */
       ┃     │                         functionalConstr20{}(
       ┃     │                             /* Spa */
       ┃     │                             \exists{testSort{}}(
       ┃     │                                 N:testSort1{},
       ┃     │                                 /* Spa */
       ┃     │                                 plain10{}(
       ┃     │                                     /* Spa */ h{}(/* D Sfa Cl */ constr00{}())
       ┃     │                                 )
       ┃     │                             ),
       ┃     │                             /* Spa */
       ┃     │                             injective11{}(
       ┃     │                                 /* Spa */
       ┃     │                                 function20MapTest{}(
       ┃     │                                     /* Fn Sfa */ \bottom{mapSort{}}(),
       ┃     │                                     /* D Sfa Cl */ constr00{}()
       ┃     │                                 )
       ┃     │                             )
       ┃     │                         )
       ┃     │                     )
       ┃     │                 ),
       ┃     │             concatList{}(
       ┃     │                 elementList{}(
       ┃     │                     /* Sfa */
       ┃     │                     /* evaluated: */
       ┃     │                     /* Spa */
       ┃     │                     functionalConstr11{}(
       ┃     │                         /* Spa */
       ┃     │                         constr10{}(
       ┃     │                             /* Spa */
       ┃     │                             injective11{}(
       ┃     │                                 /* Fn Spa */
       ┃     │                                 functional20{}(
       ┃     │                                     /* Fn Sfa */ \bottom{testSort{}}(),
       ┃     │                                     /* Fl Fn D Sfa Cl */ e{}()
       ┃     │                                 )
       ┃     │                             )
       ┃     │                         )
       ┃     │                     )
       ┃     │                 ),
       ┃     │                 elementList{}(
       ┃     │                     /*  */
       ┃     │                     \ceil{boolSort{}, testSort{}}(
       ┃     │                         /*  */
       ┃     │                         \equals{testSort1{}, boolSort{}}(
       ┃     │                             /* Spa */
       ┃     │                             \rewrites{testSort1{}}(
       ┃     │                                 /* Spa */
       ┃     │                                 \mu{}(
       ┃     │                                     x:testSort1{},
       ┃     │                                     /* Spa */
       ┃     │                                     \and{testSort1{}}(
       ┃     │                                         /* Sfa */ x:testSort1{},
       ┃     │                                         /* Fl Fn D Sfa Cl */ aSort1{}()
       ┃     │                                     )
       ┃     │                                 ),
       ┃     │                                 /* Spa */
       ┃     │                                 \floor{testSort{}, testSort1{}}(
       ┃     │                                     /* Fl Fn D Sfa Cl */ c{}()
       ┃     │                                 )
       ┃     │                             ),
       ┃     │                             /*  */
       ┃     │                             \floor{otherTopSort{}, testSort1{}}(
       ┃     │                                 /*  */
       ┃     │                                 \rewrites{otherTopSort{}}(
       ┃     │                                     /* D Sfa */ \top{otherTopSort{}}(),
       ┃     │                                     /*  */
       ┃     │                                     \equals{otherSort{}, otherTopSort{}}(
       ┃     │                                         /*  */ plain00OtherSort{}(),
       ┃     │                                         /*  */ plain00OtherSort{}()
       ┃     │                                     )
       ┃     │                                 )
       ┃     │                             )
       ┃     │                         )
       ┃     │                     )
       ┃     │                 )
       ┃     │             ))))
       ┃     │         ),
       ┃     │         /* Fl Fn D Sfa Cl */ /* InternalList: */ unitList{}()
       ┃     │     )
       ┃     │ has resulted in \bottom.
```

**Comment by ttuegel** [comment]
The core of this investigation seems to be this term:

```
plain10{}( \bottom{testSort{}}() )
```

[`Kore.Step.Simplification.Pattern.simplify`](https://github.com/kframework/kore/blob/21959d95b33d7306d6e80cc0d32d1158aea06bbc/kore/src/Kore/Step/Simplification/Pattern.hs#L70) should turn this to `\bottom{testSort{}}()`.

**Comment by ttuegel** [comment]
This will be fixed by #2468.
